File: Week9/51.js
var solveNQueens = function(n) {
    var res=[]
    var s=new Array(n).join(".")
    var slove=function(row,temp,cols,xs,ys){
        if (row==n){
            res.push(temp)
        }else{
            for(let j=0;j<n;j++){
                if ((!cols.has(j)) && (!xs.has(row-j)) && (!ys.has(row+j))){
                    var cur=`${s.slice(0,j)}Q${s.slice(j)}`
                    temp.push(cur)
                    cols.add(j)
                    xs.add(row-j)
                    ys.add(row+j)
                    slove(row+1,Array.from(temp),cols,xs,ys)
                    temp.pop()
                    cols.delete(j)
                    xs.delete(row-j)
                    ys.delete(row+j)
                }
            }
        }
    }
    slove(0,[],new Set(),new Set(),new Set())
    return res
};

File: Week10/51.py
import math
class Solution:
    def solveNQueens(self, n):
        self.res=[]
        self.s=[["."]*n  for  _ in range(n)]
        self.n=n
        self.size=(1<<n)-1
        self.solve(self.s[:],0,0,0,0)
        return self.res
    def solve(self,s,i,ns,xs,ys):
        if i>=self.n:
            self.res.append(["".join(t) for t in s])
            return
        can=self.size&(~(ns|xs|ys))
        while can!=0:
            temp=can&(-can)
            j=int(math.log2(temp))
            s[i][j]="Q"
            self.solve(s[:],i+1,ns|temp,(xs|temp)<<1,(ys|temp)>>1)
            s[i][j]="."
            can&=(can-1)